examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,839 @@
+#include "common.h"
+#include "llama.h"
+#include "ggml.h"
+
+#ifdef GGML_USE_CUDA
+#include "ggml-cuda.h"
+#endif
+
+#ifdef GGML_USE_METAL
+#include "ggml-metal.h"
+#endif
+
+#include <cstdio>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <fstream>
+
+#define DEBUG_POS 2
+
+// TODO read everything over and make sure it makes sense because I'm dropping logic errors left and right - Christian
+
+struct callback_data {
+    std::vector<uint8_t> data;
+    ggml_context * ctx_ggml;
```

Comment:
Also pay attention that, we should prevent alloc a new `ggml_context` too many times. We can probably have 2 ggml_context(s) in our program:
- One for getting the embeddings from cb_eval and calculate diff
- One for PCA

To be honest, I also don't know what's the best way to calculate the mem_size needed for create the context. @christianazinn you can pick whatever works for you (even more than needed), then we will see at later stage how to calculate it.

Reply:
Another hack would be to set `no_alloc` to true. This will prevent alloc `tensor->data` inside the given ggml_context. We can then manually alloc it via `malloc` (see me `cb_eval` for example)

Reply:
`no_alloc=true` is probably the best way to handle cb_eval `ggml_context`. PCA power iteration context is easy - we just need a square matrix and vector of dimensions `cb_data.n_embd`, which is constant. However the `n_tokens` varies between prompts so we would need to keep scaling the `ggml_context` used to store pos/neg/diff tensors. I'll have to think about how to do this, but first, we can definitely just keep one `ggml_context` for PCA.